File: src/transformers/features/reflect/ReflectMetadataTransformer.ts
import ts from "typescript";

import { LiteralFactory } from "../../../factories/LiteralFactory";
import { MetadataCollection } from "../../../factories/MetadataCollection";
import { MetadataFactory } from "../../../factories/MetadataFactory";

import { IMetadataApplication } from "../../../schemas/metadata/IMetadataApplication";
import { Metadata } from "../../../schemas/metadata/Metadata";

import { ITransformProps } from "../../ITransformProps";
import { TransformerError } from "../../TransformerError";

export namespace ReflectMetadataTransformer {
  export const transform = (
    props: Pick<ITransformProps, "context" | "expression">,
  ): ts.Expression => {
    if (!props.expression.typeArguments?.length)
      throw new TransformerError({
        code: "typia.reflect.metadata",
        message: "no generic argument.",
      });

    // VALIDATE TUPLE ARGUMENTS
    const top: ts.Node = props.expression.typeArguments[0]!;
    if (!ts.isTupleTypeNode(top)) return props.expression;
    else if (top.elements.some((child) => !ts.isTypeNode(child)))
      return props.expression;

    // GET TYPES
    const types: ts.Type[] = top.elements.map((child) =>
      props.context.checker.getTypeFromTypeNode(child as ts.TypeNode),
    );
    if (types.some((t) => t.isTypeParameter()))
      throw new TransformerError({
        code: "typia.reflect.metadata",
        message: "non-specified generic argument(s).",
      });

    // METADATA
    const collection: MetadataCollection = new MetadataCollection();
    const metadatas: Array<Metadata> = types.map((type) => {
      const result = MetadataFactory.analyze({
        checker: props.context.checker,
        transformer: props.context.transformer,
        options: {
          escape: true,
          constant: true,
          absorb: true,
          functional: true,
        },
        collection,
        type,
      });
      if (result.success === false)
        throw TransformerError.from({
          code: "typia.reflect.metadata",
          errors: result.errors,
        });
      return result.data;
    });

    // CONVERT TO PRIMITIVE TYPE
    const app: IMetadataApplication = {
      metadatas: metadatas.map((metadata) => metadata.toJSON()),
      components: collection.toJSON(),
    };
    return LiteralFactory.write(app);
  };
}

File: packages/typia/native/transform/features/reflect/ReflectMetadataTransformer.go
package reflect

import (
  shimast "github.com/microsoft/typescript-go/shim/ast"
  nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
  schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
  nativetransform "github.com/samchon/typia/packages/typia/native/transform/internal"
)

type reflectMetadataTransformerNamespace struct{}

var ReflectMetadataTransformer = reflectMetadataTransformerNamespace{}

func (reflectMetadataTransformerNamespace) Transform(props nativetransform.ITransformProps) *shimast.Node {
  if props.Expression == nil || props.Expression.TypeArguments == nil || len(props.Expression.TypeArguments.Nodes) == 0 {
    panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
      Code:    "typia.reflect.metadata",
      Message: "no generic argument.",
    }))
  }

  top := props.Expression.TypeArguments.Nodes[0]
  elements := reflectTransformer_tupleElements(top)
  if elements == nil {
    return props.Expression.AsNode()
  }

  types := make([]*shimast.Node, 0, len(elements))
  for _, child := range elements {
    if child == nil {
      return props.Expression.AsNode()
    }
    types = append(types, child)
  }

  components := schemametadata.NewMetadataCollection()
  schemas := make([]any, 0, len(types))
  for _, node := range types {
    typ := props.Context.Checker.GetTypeFromTypeNode(node)
    if typ != nil && typ.IsTypeParameter() {
      panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
        Code:    "typia.reflect.metadata",
        Message: "non-specified generic argument(s).",
      }))
    }
    result := nativefactories.MetadataFactory.Analyze(nativefactories.MetadataFactory_IProps{
      Checker:     props.Context.Checker,
      Transformer: props.Context.Transformer,
      Options: nativefactories.MetadataFactory_IOptions{
        Escape:     true,
        Constant:   true,
        Absorb:     true,
        Functional: true,
      },
      Components: components,
      Type:       typ,
    })
    if result.Success == false {
      panic(nativetransform.TransformerError_from(struct {
        Code   string
        Errors []nativetransform.TransformerError_MetadataFactory_IError
      }{
        Code:   "typia.reflect.metadata",
        Errors: reflectTransformer_errors(result.Errors),
      }))
    }
    schemas = append(schemas, result.Data.ToJSON())
  }

  return reflectTransformer_literal(map[string]any{
    "schemas":    schemas,
    "components": components.ToJSON(),
  })
}
